### DISCUSSION TITLE
Is there a Guide to Adding functions ?

### DISCUSSION THREAD

**Discussion by JonTidswell** [shibatch/sleef#410] [Q&A]
I was looking to add some functions to sleef.

To simplify testing, I would like to get the scalar versions working before writing simd versions.

Is there an guide to incrementally add functions to the dispatcher ?

thanks,
  -- JonT

**Comment by shibatch** [comment]
Generally speaking, adding a function that meets the accuracy requirements is pretty hard.
When I add a new function to SLEEF, I first implement it as a scalar function, and test its accuracy with a small program independent of SLEEF.

If you are planning to upstream your function, I would like to see the scalar implementation first.

**Reply by JonTidswell** [reply]
I am only targeting _faithful_ not _correct_ rounding, because well ... the table makers dilemma. :wink:

Some coefficients I found using a linear program, and the remainder I used [sollya](https://www.sollya.org/)'s `fpminimax` to derive them.

Then I've built some Julia workbooks to try to show that using fma and with **key** pieces in double-double that
1. the floating point implementations are under 0.5 ulps  so that I can be confident that they are under 1 ulp after final rounding. (in excess of $10^8$ samples under 0.4 ulps is not a proof, but its a start)
2. the strength reduction and reconstruction holds on the input domains on which I am interested.

Notes:
1. At this stage I'm planning to inline some simpler functions to exploit pipeline parallelism, but I should make use of whole function inlining [#282] rather than do it manually. 
2. I know that depending on the definition of `ulp` that 1 ulp is not necessarily equivalent to faithful rounding at binade boundaries, so I will have to check them explicitly. (Not done yet.)

Since I was hoping to publish the key pieces, I'd rather not discuss the function specifics in an open forum.

thanks,
-- JonT

**Comment by shibatch** [comment]
Sounds good, but how can I help you?

**Reply by JonTidswell** [reply]
How about I try to create a worked example using the erfinv function ?
$$\mathrm{erfinv}\left( x \right) = \mathrm{erf}^{-1}\left( x \right)$$

Its a [well studied problem](http://www.mimirgames.com/articles/programming/approximations-of-the-inverse-error-function/), sometimes used in Monte-Carlo simulations, with good public algorithms.

I think I will get stuck in an equivalent position -- testing when I have only the scalar implementation, and no dispatcher.

**Reply by shibatch** [reply]
Good. Please go ahead.

**Reply by JonTidswell** [reply]
So I've created `xerfinv_u4`, `xerfinvf_u4` in sleefsp.c/sleefdp.c by cut-n-paste from Giles published cuda code, and added some conditional assignments for boundary conditions.
Testing equivalent code in Julia showed errors between 3.5 and 4 ulps, hence the name choice.
I created a function `mpfr_erfinv` as a test helper to simulate the missing mpfr function.
I've also copy-created some basic tests in tester.c.

So now Im [at the point](https://github.com/JonTidswell/sleef/tree/erfinv) where I'm missing the guide as to how it connect all the pieces properly.

Im hoping this illustrates where I'm getting stuck.

**Reply by shibatch** [reply]
Good. Please rewrite those `if` statements to ternary operators like in xerfc_u15.
You have to remove all `if` statements.

**Reply by JonTidswell** [reply]
Warp divergence is part of Giles algorithm (its very low probability: IIRC ~ 0.5 %).
In a pure SIMD architecture (SSE/AVX/etc), this obviously needs to be coded differently but such an environment would provide the instructions (or intrinsics) for testing all data points in parallel; this capability is not in pure C; but it is not pure SIMD, so its not an obvious problem.
[ Edit: I've added some extra variables to better reflect the requirements, so it should translate more directly into pure SIMD code. ]

**Reply by shibatch** [reply]
But anyway you need the slow path to compute the full range of input domain.
Or are you going to evaluate two polynomials in turn in the slow path?

**Reply by JonTidswell** [reply]
Giles design is slow path evaluates both polynomials.

**Reply by shibatch** [reply]
One more thing to consider is how the polynomial will be evaluated. Your current code is with Horner, but many other functions use Estrin, which is faster but less accurate.

**Reply by shibatch** [reply]
Can you regenerate the coefficients?

**Reply by JonTidswell** [reply]
Giles definition is Horner. 
But your implied criticism is well placed: actually implementing two separate paths and also not using an Estrin scheme, would be a poor implementation.
Conceptually I could generate new coefficients, but if I was going to go to that effort I would make a larger algorithm change and not use the erfinv function.

**Reply by JonTidswell** [reply]
Given the intentionally strongly weighted probability of using the first polynomial, implementing them in parallel using pipelining would not gain much, so I will revisit the use of Estrin and evaluate what it does to the accuracy.

**Reply by shibatch** [reply]
I don’t understand. If only the fast path matters, it will be like two times faster with Estrin.
Anyway, I will check your implementation tomorrow.

**Reply by shibatch** [reply]
There seems something wrong with your mpfr_erfinv.
It refers to `ret` before substituting it.

https://github.com/JonTidswell/sleef/blob/3b5ac85214271ab9eea421d136787a92c00b4b6d/src/libm-tester/testerutil.c#L447

**Reply by shibatch** [reply]
Assuming that that `ret` is actually `arg`, the mpfr_erfinv still does not return accurate results.
For example, erfinv(0.5) is 0.47693627620446987338... according to wolfram alpha,
but your function returns 0.47694314105631873... with 1280-bit precision setting.

**Reply by JonTidswell** [reply]
Ooops, editing error. Sorry. It should be `tmp1`.

I did some more testing, it now matches my Julia code.

And for values that translate to Wolfram well (ie converting decimal fractions to Wolfram introduces differences):

```
Wolfram: 2^-30 :       8.2536314190622234144278198360757195830466198505176083048185 × 10^-10
test code: 0x1p-30 : +0.825363141906222341442781983607571958302028017499959e-9

Wolfram: 2^-10 :   0.0008654561979671375534512120376117105096279820389742517983453693
test code: 0x1o-10 : +0.865456197967137553451212037611710509626809948605394e-3

Wolfram: 2^-2 :     0.2253120550121781047250140139522775547821184478072467576007828949
test code 0x1p-2 : +0.225312055012178104725014013952277554782361895851697e+0

Wolfram: 1/2 :    0.4769362762044698733814183536431305598089697490594706447038826959
test code 0.5 :  +0.476936276204469873381418353643130559808854836858638e+0
```

**Reply by shibatch** [reply]
Good.
But, the error of the function seems to sometimes exceed 4 ulps.

arg = +0.1400544893812095959
test +0.12476441516943169441 0x1.ff08f8adb7667p-4
corr +0.12476441516943174992 0x1.ff08f8adb766bp-4

**Reply by JonTidswell** [reply]
Mmmm.
It is more than 4.
Was this generated with Horner or changes to Estrin ? (I renamed them to u5 with Estrin because I found examples over 4 with Estrin in single precision, and too close to 4 in double precision to believe that they would stay under. )

**Reply by JonTidswell** [reply]
Its Horner, and this example goes away with Estrin.

**Reply by shibatch** [reply]
Could you use POLY21 macro, which is defined in src/common/estrin.h?
The last few terms should be evaluated with Horner.
In other functions, the last few terms are evaluated in DD precision to make the resulting accuracy under 1ulp.

**Reply by JonTidswell** [reply]
Probably, but **I** start to run into the problem that this thread is about - how to properly integrate and test functions in sleef.
( Without the macro I can test easily enough in Julia. :)

[ For reference I think a test value of 0.14 has a 2 ulp error when the polynomial is evaluated in mpfr. ]

Wolfram: erfinv( 0.14000000000000001332267629550187848508358001708984375 )
  * 0.12471536794266059546900356708549725509490189605721092

xerfinv ( 0.14 ) evauluated using arbitrary precision
  * 0.12471536794266056692663257298785432210775298235778314031618117312126

Error is ~ 2.7755575615628914e-17
ulp is ~ 1.3877787807814457e-17

**Reply by shibatch** [reply]
At this point, I use a simple program separate from sleef to check the accuracy, but you can carry out continuous testing with tester2.
